aioshelly/rpc_device.py

```diff
@@ -0,0 +1,202 @@
+"""Shelly Gen2 RPC based device."""
+from __future__ import annotations
+
+from typing import Any, Dict
+
+import aiohttp
+
+from .common import ConnectionOptions, get_info
+from .exceptions import AuthRequired, NotInitialized
+from .wsrpc import WsRPC
+
+
+def mergedicts(dict1, dict2):
+    """Deep dicts merge."""
+    for k in set(dict1.keys()).union(dict2.keys()):
+        if k in dict1 and k in dict2:
+            if isinstance(dict1[k], dict) and isinstance(dict2[k], dict):
+                yield (k, dict(mergedicts(dict1[k], dict2[k])))
+            else:
+                yield (k, dict2[k])
+        elif k in dict1:
+            yield (k, dict1[k])
+        else:
+            yield (k, dict2[k])
+
+
+class RpcDevice:
+    """Shelly RPC device reppresentation."""
+
+    def __init__(
+        self,
+        aiohttp_session: aiohttp.ClientSession,
+        options: ConnectionOptions,
+    ):
+        """Device init."""
+        self.aiohttp_session = aiohttp_session
+        self.options = options
+        self.shelly = None
+        self._status: Dict[str, Any] | None = None
+        self._event: Dict[str, Any] | None = None
+        self._device_info = None
+        self._config = None
+        self._wsrpc = WsRPC(options.ip_address, self._on_notification)
+        self._update_listener = None
+        self._initialized = False
+        self._initializing = False
+
+    @classmethod
+    async def create(
+        cls,
+        aiohttp_session: aiohttp.ClientSession,
+        options: ConnectionOptions,
+        initialize: bool = True,
+    ):
+        """Device creation."""
+        instance = cls(aiohttp_session, options)
+
+        if initialize:
+            await instance.initialize()
+
+        return instance
+
+    async def _on_notification(self, method, params):
+        if method == "NotifyStatus" and params is not None:
+            self._status = dict(mergedicts(self._status, params))
+        elif method == "NotifyEvent" and params is not None:
+            self._event = params
+
+        if self._update_listener:
+            self._update_listener(self)
+
+    @property
+    def ip_address(self):
+        """Device ip address."""
+        return self.options.ip_address
+
+    async def initialize(self):
+        """Device initialization."""
+        self._initializing = True
+        self._initialized = False
+        try:
+            await self._wsrpc.connect(self.aiohttp_session)
+            self.shelly = await get_info(self.aiohttp_session, self.options.ip_address)
+
+            if self.options.auth or not self.shelly["auth_en"]:
+                await self.update_device_info()
+                await self.update_config()
+                await self.update_status()
```

Comment:
Should we do it in parallel? We used to not do that because of COAP response issues, but we don't have that here? 
```suggestion
                await asyncio.gather(
                    self.update_device_info(),
                    self.update_config(),
                    self.update_status(),
                )
```

Reply:
Yes that will work